lib/agent0/agent0/hyperdrive/exec/run_agents.py

```diff
@@ -380,3 +391,99 @@ def build_wallet_positions_from_data(
         longs=long_obj,
         shorts=short_obj,
     )
+
+
+def build_wallet_positions_from_chain(
+    wallet_addr: str, hyperdrive_contract: IERC4626HyperdriveContract, base_contract: ERC20MintableContract
+) -> HyperdriveWallet:
+    """Builds a wallet position based on gathered data.
+
+    Arguments
+    ---------
+    wallet_addr: str
+        The checksum wallet address
+    hyperdrive_contract: Contract
+        The Hyperdrive contract to query the data from
+    base_contract: Contract
+        The base contract to query the base amount from
+
+    Returns
+    -------
+    HyperdriveWallet
+        The wallet object build from the provided data
+    """
+    # Contract call to get base balance
+    base_amount: int = base_contract.functions.balanceOf(wallet_addr).call()
+    # TODO do we need to do error checking here?
+    base_obj = Quantity(amount=FixedPoint(scaled_value=base_amount), unit=TokenType.BASE)
+
+    # Contract call to get lp balance
+    asset_id = encode_asset_id(AssetIdPrefix.LP, 0)
+    lp_amount: int = hyperdrive_contract.functions.balanceOf(asset_id, wallet_addr).call()
+    lp_obj = FixedPoint(scaled_value=lp_amount)
+
+    # Contract call to get withdrawal positions
+    asset_id = encode_asset_id(AssetIdPrefix.WITHDRAWAL_SHARE, 0)
+    withdraw_amount: int = hyperdrive_contract.functions.balanceOf(asset_id, wallet_addr).call()
+    withdraw_obj = FixedPoint(scaled_value=withdraw_amount)
+
+    # We need to gather all longs and shorts from events
+    # and rebuild the current long/short positions
+    # Open Longs
+    open_long_events = hyperdrive_contract.events.OpenLong.get_logs(fromBlock=0)
+    long_obj: dict[int, Long] = {}
+    for event in open_long_events:
+        maturity_time = event["args"]["maturityTime"]
+        long_amount = FixedPoint(scaled_value=event["args"]["bondAmount"])
+        # Add this balance to the wallet if it exists, create the long object if not
+        if maturity_time in long_obj:
+            long_obj[maturity_time].balance += long_amount
+        else:
+            long_obj[maturity_time] = Long(balance=long_amount, maturity_time=maturity_time)
+    # Close Longs
+    close_long_events = hyperdrive_contract.events.CloseLong.get_logs(fromBlock=0)
+    for event in close_long_events:
+        maturity_time = event["args"]["maturityTime"]
+        long_amount = FixedPoint(scaled_value=event["args"]["bondAmount"])
+        assert maturity_time in long_obj
```

Comment:
can you add error messages to these asserts? like "wallet had close event without corresponding open event" and "wallet had negative long balance, which is not supported"